### ISSUE TITLE
Proposal to discard kernel .fartext.init section after startup

### ISSUE THREAD

**Issue by ghaerr** [ghaerr/elks#1644]
Hello @tkchia,

When the current kernel is built with CONFIG_FARTEXT_KERNEL=y, two labeled portions of kernel code are treated differently: `INITPROC` for kernel initialization routines, used only once, and `FATPROC` for the FAT filesystem. Both are used to reduce the size of the near code segment, allowing for fast kernel execution most of the time and a larger than 64K total code segment.

The `INITPROC` macro is defined as:
```
#define INITPROC __far __attribute__ ((far_section, noinline, section (".fartext.init")))
```
I am considering the idea of discarding the `.fartext.init` section after kernel initialization, and adding that memory back into the main memory free list, which would then allow for use by application program code or data segments, or external kernel buffers.

In order to do this, the `elks-small.ld` linker script will need some adjustment. I would like to ask your help on this, if possible. What is needed would be to:
- Sort the `.fartext` sections, and place the combined `.fartext.init` as the final section in the kernel code segment. Since the current system physical memory layout is the kernel data segment followed by the kernel code segment(s), this would put the `.fartext.init` section at the highest memory address used, which with some other consideration, would allow it to likely be merged into a larger block of available memory.
- Declare two global symbols, accessible after linking from the kernel, indicating the starting segment and size of the `.fartext.init` section.

With the above, it would be quite easy to just add the unused INITPROC section after initialization to the memory free list. With a little more work, other kernel main memory (e.g. buffer, etc) allocation could be delayed to allow for a contiguous addition to the free list, as described above.

Do you have some suggestions as to how the `elks-small.ld` could be modified for this? Although the same script is used for the kernel and applications programs, I don't think this will be an issue.

Thank you!

**Comment by tkchia** [comment]
Hello @ghaerr,

A while back I crafted a linker script for `newlib-ia16` to allow building of MS-DOS tiny-model terminate-and-stay-resident programs (https://github.com/tkchia/newlib-ia16/blob/bad4dfe895de09aae26271de18acfa99ca286a71/libgloss/ia16/dtr-mt.ld.in); we could probably employ a similar method.

Basically the script tries to put in the `.text` section only those `text.`... sections that are _not_ named `.text.startup` or `.text.startup.*` or `.text.exit` or `.text.exit.*`.  (This is done using some moderately ugly wildcards.  It seems the linker script syntax does not provide an easy way to exclude input sections with particular names &mdash; there is an `EXCLUDE_FILE` facility, but that is it.)
```
SECTIONS
  {
	...
	.text : {
		__stext_keep = .;
		*(.startupA)
		*(.head) *(.head.*)
		*(.startupB)
		*(.text) *(.text. .text.[^se]* .text.s[^t]* .text.st[^a]*
			   .text.sta[^r]* .text.star[^t]* .text.start[^u]*
			   .text.startu[^p]* .text.startup[^.]* .text.e[^x]*
			   .text.ex[^i]* .text.exi[^t] .text.exit[^.]*)
		__etext_keep = .;
	}
	...
	.text.startup ALIGN (__heap_end_minimum + __heaplen_val
					    + __stklen_val_nokeep, 2)
		  : AT (__stext_nokeep_load)
	{
		__stext_nokeep = .;
		*(.startupC)
		__stack_end_nokeep = . & -0x10;
		*(.msdos_init) *(.msdos_init.*)
		...
		*(.text.startup) *(.text.startup.*)
	} ...
}
```
Thank you!

**Comment by ghaerr** [comment]
Hello @tkchia,

Thank you for the linker script descriptions, I can see it's quite messy! I'll give it a shot, with the appropriate wildcard changes for `.fartext.init` (or possibly just rename `.fartext.init` to `.fartext.start` for the time being).

I do notice that from the kernel mapfile, it seems that `.fartext.*` isn't sorted, can I just insert a `SORT(".fartext.*")` previous to the section statement `*(.fartext .fartext$* ".fartext.*[^&]")`, which include the actual segment contents, or is no sort needed because the entire section would be excluded using the mechanisms you described above?

Thank you!

**Comment by ghaerr** [comment]
Hello @tkchia,

Please ignore my previous question above.

As a basic experiment, I added the following couple of lines in `elks-small.ld`:
```
   .fartext 0x20000 : {
        *(".fartext!*" ".fartext.*!")

        __start_fartext = .; // <--- added this
        /* Actual segment contents. */
        *(.fartext .fartext$* ".fartext.*[^&]")
        __end_fartext = .; // <- added this
```
as well as the following in the kernel:
```
    extern void INITPROC __start_fartext();
    printk("addr %08lx\n", __start_fartext);
```
However, due to the special medium model segment/linear address conversions taking place in the linker, I'm getting the following link error:
```
init/main.o:function kernel_banner: error: undefined reference to '__start_fartext!'
```
I've tried adding the line "__start_fartext! = .` without success, not understanding how to declare a missing segment start marker.

[EDIT]: I have eliminated the undefined symbol reference error by declaring an empty function using:
```
void INITPROC __start_fartext(void) {}
```
and the linker will assign the address of the `__start_fartext` function to the start of section `.fartext.init` as desired according to the linker script below. The linker script now looks like this, using your suggestions:
```
.fartext 0x20000 : {      
        *(".fartext!*" ".fartext.*!")

        /* Actual segment contents. */
        *(.fartext .fartext.
            ".fartext.[^i]*" ".fartext.i[^n]*" ".fartext.in[^i]*" ".fartext.ini[^t]*")
        __start_fartext = .;
        *(.fartext .fartext.init ".fartext.init")
        __end_fartext = .;

        /* IA-16 segment end markers. */
        *(".fartext&*" ".fartext.*&")
        . = ALIGN(0x10);
}
```
Does that look like it includes all symbols, including the special "!" and "&" symbols? I'll post a PR for your final inspection.

Thank you!
